ValueRange analysis for indirect indexing

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105055
* __->__ #102611
* #104987
* #104975

We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

lezcano committed:
ValueRange analysis for indirect indexing

We do so by forwarding ValueRange analysis from IR buffers to CSEvars

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/common.py:
```diff
@@ -777,8 +797,34 @@ class CSEProxy:
             def __getattr__(name):
                 def inner(*args, **kwargs):
+                    buf_bounds = ValueRanges.unknown()
+                    if name == "constant":
```
Note to self: This shouldnt' be necessary. Not sure why was it not being found in the bounds. I'll look into it.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1136,9 +1136,27 @@ def __init__(self, line, var, mask, size_map):
+                size, size_str = self.size_map[(self.var, self.mask)]
+
+                assert_min = var.bounds.lower < 0
+                assert_max = var.bounds.upper > size
```
Does this work when `size` is an expression?

lezcano replied:
nah, in the dynamic shapes case we should always generate asserts. I have been working on getting the other three PRs to work. Now those are passing CI, I'll have a look at this one

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano commented:
@eellison @peterbell10 this is ready for review. This PR elides the device_asserts in most cases. 

A limitation it still has (that was also present in the initial index strenght PR) is that it does not transfer bounds information between loops. As such, for example, we can change `_unsafe_index` in `upsample_bilinear` ops, and the asserts would be elided, but since we don't pass the bounds to the backward, we still get `assert`s in the backward.

lezcano commented:
Also, once a PR similar to https://github.com/pytorch/pytorch/pull/100967/ is merged, whoever implements it will need to make sure that the bounds are properly propagated on the auxiliary variables created on `cpp.py`.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1161,11 +1162,27 @@ def __init__(self, line, var, mask, size_map):
-                cond_print = f"0 <= {self.var} < {size}"
+                size, size_str = self.size_map[(self.var, self.mask)]
+
+                assert_min = (var.bounds.lower < 0) == sympy.true
+                assert_max = (var.bounds.upper > size) == sympy.true
```
I take it this is `sympy.Lt` for dynamic shapes, so `sympy.Lt != sympy.true` and we generate an assert?

```suggestion
                assert_min = (self.var.bounds.lower < 0) == sympy.true
                assert_max = (self.var.bounds.upper >= size) == sympy.true
```

peterbell10 replied:
I'm assuming `bound.upper` is a closed range `[lower, upper]` but if it's half open then `> size` is correct.

lezcano replied:
https://github.com/pytorch/pytorch/pull/104987. Funnily enough, this issue has been there since the inception of these bounds. How about that.

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -6784,6 +6786,48 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+        def test_optimize_indexing_assert(self):
+            def fn(x: torch.Tensor) -> torch.Tensor:
+                s = 0.7 * torch.arange(x.shape[0], device=x.device)
+                return x[s.long()]
```
Do we have some test cases where the assert isn't removed

lezcano replied:
other than those where we pass bounded tensors from one kernel to another, of course.

peterbell10 replied:
@lezcano I think it's important to test that bounds are generated in some cases, like the `index_put_` example where you were getting no upper bound generated.

peterbell10 replied:
Less important, by another good test would be user supplied indices with a clamp call that forces it into range, which happens in `grid_sampler` for example.

lezcano replied:
Added a test for `index_put_`. The `clamp(0)` example does not work yet, I will add a PR to this stack fixing it.

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -6784,6 +6786,48 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+                            msg=f"Indirect indexing not present in code:\n{code}",
+                        )
+            # ...but the assert is now there
+            self.assertTrue("device_assert" in code)
```
Because we consider all the dynamic shapes unbounded.

peterbell10 replied:
So it doesn't have a symbolic bound where we could say that `0.7 * (s0 - 1) < s0` is true?

lezcano replied:
Yeah, the analysis could be improved for dynamic shapes. This was put in place to unblock the landing of dynamic_shapes=True.

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -6784,6 +6786,48 @@ def fn(x: torch.Tensor) -> torch.Tensor:
+
+            # If we happen to have dynamic shapes
+            x = torch.randn(9, device="cuda")
+            code = run_and_get_triton_code(fn_opt, x, _reset_dynamo=False)
```
Wouldn't it be clearer to compile with `dynamic=True`?

lezcano replied:
sure thing, let me do that.

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -731,13 +742,16 @@ def __init__(self, args=None):
     def set_current_node(self, node):
         prior = self.current_node
         self.current_node = node
+        self.bounds = node._body.bounds().get_bounds()
```
just curious, why set bounds here and not on initialization ?

lezcano replied:
Because this is run on LoopBody, that is, on each node within a SchedulerNode. This was already the case in your initial implementation.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -805,8 +819,21 @@ class CSEProxy:
+                        fx_node = V.interpreter.current_node
+                        buf_bounds = self.bounds.get(fx_node, ValueRanges.unknown())
+                        # Sanity check: The variable is either bounded or unbounded
+                        assert (
+                            buf_bounds is not None
+                            or fx_node in self.bounds.unbounded_vars
```
commented this earlier, but it's still a little confusing to me that we could have both `ValueRanges.unknown()` or `None` when both are semantically equivalent. i think it would be clearer to have just one of them

lezcano replied:
removing the assert, as it was a bit stale, given that we were doing a `get` with a `ValueRanges.unknown()` before.

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -805,8 +819,21 @@ class CSEProxy:
+                    buf_bounds = ValueRanges.unknown()
+                    if hasattr(V.interpreter, "current_node"):
+                        fx_node = V.interpreter.current_node
+                        buf_bounds = self.bounds.get(fx_node, ValueRanges.unknown())
```
worth turning `bounds.get(fx_node, ValueRanges.unknown())` into an API ?

lezcano replied:
I think not yet. This may be interesting to discuss if we lift the analysis to the whole graph (and the backward graph). At the moment `self.bounds` is not the main entrypoint. `csevar.bounds` is.

eellison commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -83,6 +83,19 @@ def __contains__(self, x):
         x = simple_sympify(x)
         return sympy_generic_le(self.lower, x) and sympy_generic_le(x, self.upper)
 
+    def tighten(self, other):
```
add docstring/type annotation ?

eellison commented on torch/_inductor/codegen/triton.py:
```diff
@@ -95,8 +95,8 @@ def triton_constant(value):
-    def __init__(self, name):
-        super().__init__(name)
+    def __init__(self, name, bounds):
+        super().__init__(name, bounds)
```
add type annotation ?

eellison commented on torch/_inductor/codegen/common.py:
```diff
@@ -805,8 +819,21 @@ class CSEProxy:
+                    buf_bounds = ValueRanges.unknown()
+                    if hasattr(V.interpreter, "current_node"):
+                        fx_node = V.interpreter.current_node
+                        buf_bounds = self.bounds.get(fx_node, ValueRanges.unknown())
```
in some of the other places `self.bounds` is a `ValueRanges` object and in some places `self.bounds` is a dict.. maybe using `value_range` for the instance would be clearer. or type annotations.

lezcano replied:
Changed name to `node_to_bounds` and added type annotations.

eellison reviewed (commented):
Cool ! Is it always valid to tighten the bounds of an expression as we are doing in CSE ?  For example:

```
def foo(x):
    if x < 0:
        # x bound [-inf, 0)
        y = x + x
    else:
        # x bound [0, inf]
        y = x * 2
```

maybe we don't have any path dependent bounds analysis right now so it this doesn't matter.

lezcano commented:
Note that those two values, even if they have the same name in Python, they would have different names in the IR, so that shouldn't be a problem. Every CSE variable will refer to the same physical variable in triton/C++, so if we have two different bounds for the variable, we should take their intersection.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano commented:
Addressed all the reviews.

eellison reviewed (approved):
🚀 🚀

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1171,11 +1173,27 @@ def __init__(self, line, var, mask, size_map):
-                cond_print = f"0 <= {self.var} < {size}"
+                size, size_str = self.size_map[(self.var, self.mask)]
+
+                assert_min = (var.bounds.lower < 0) == sympy.true
+                assert_max = (var.bounds.upper >= size) == sympy.true
```
This is capturing `var` from the parent scope, you want `self.var`.
```suggestion
                assert_min = (self.var.bounds.lower < 0) == sympy.true
                assert_max = (self.var.bounds.upper >= size) == sympy.true
```

lezcano replied:
missed that one. Fixed now.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1171,11 +1173,28 @@ def __init__(self, line, var, mask, size_map):
+                size, size_str = self.size_map[(self.var, self.mask)]
+
+                # We assert if we've not been able to prove the bound
+                assert_min = not (self.var.bounds.lower >= 0)
+                assert_max = not (self.var.bounds.upper < size)
```
How does this work? If I try this locally with a sympy Expression I get
```
TypeError: cannot determine truth value of Relational
```

lezcano replied:
Yep, I thought it was equivalent but it's not even equivalent. Now I argue that the two previous options were wrong. Now I think I got the condition right.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -750,8 +752,8 @@ def __init__(
-        self.indirect_max_sizes_expr = {}  # Upper bounds for indirect_indexing
-        self.indirect_max_sizes_printed = {}  # Upper bounds, printed as a string
+        # Upper bounds for indirect_indexing and their str representation
+        self.indirect_max_sizes: Dict[Tuple[str, str] : [sympy.Expr, str]] = {}
```
Typo?
```suggestion
        self.indirect_max_sizes: Dict[Tuple[str, str], [sympy.Expr, str]] = {}
```

peterbell10 reviewed (approved):
Small typo, otherwise looks good.

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "ValueRange analysis for indirect indexing"


We do so by forwarding ValueRange analysis from IR buffers to CSEvars

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
